### ISSUE TITLE
Picket fence observations are not handled as intended

### ISSUE THREAD

**Issue by d3v-null** [MWATelescope/mwa_hyperdrive#17]
This [warning](https://github.com/MWATelescope/mwa_hyperdrive/blob/main/src/calibrate/params/mod.rs#L854) implies that the intended behavior for picket fence observations is to calibrate only the first contiguous band, however this is not what happens when raw data is provided. Instead, Hyperdrive crashes.

```
thread '<unnamed>' panicked at 'ndarray: index (0, 111) is out of bounds for array of shape [7626, 111]', src/vis_io/read/raw/mod.rs:659:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: mwa_hyperdrive::vis_io::read::raw::RawDataReader::read_inner::{{closure}}::{{closure}}
   3: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:784:29
   4: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:106:27
   5: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:84:21
   6: core::iter::adapters::filter::filter_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/filter.rs:37:44
   7: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:106:27
   8: core::iter::traits::iterator::Iterator::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:2370:21
   9: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:112:9
  10: <core::iter::adapters::filter::Filter<I,P> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/filter.rs:101:9
  11: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:124:9
  12: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:112:9
  13: core::iter::traits::iterator::Iterator::for_each
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:787:9
  14: mwa_hyperdrive::vis_io::read::raw::RawDataReader::read_inner::{{closure}}
             at ./src/vis_io/read/raw/mod.rs:643:21
  15: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:784:29
  16: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:106:27
  17: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:84:21
  18: core::iter::adapters::filter::filter_fold::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/filter.rs:37:44
  19: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:106:27
  20: core::iter::traits::iterator::Iterator::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:2370:21
  21: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:112:9
  22: <core::iter::adapters::filter::Filter<I,P> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/filter.rs:101:9
  23: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/map.rs:124:9
  24: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/adapters/enumerate.rs:112:9
  25: core::iter::traits::iterator::Iterator::for_each
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/iter/traits/iterator.rs:787:9
  26: mwa_hyperdrive::vis_io::read::raw::RawDataReader::read_inner
             at ./src/vis_io/read/raw/mod.rs:633:13
  27: <mwa_hyperdrive::vis_io::read::raw::RawDataReader as mwa_hyperdrive::vis_io::read::VisRead>::read_crosses
             at ./src/vis_io/read/raw/mod.rs:759:9
  28: mwa_hyperdrive::calibrate::params::CalibrateParams::read_crosses
             at ./src/calibrate/params/mod.rs:1204:9
  29: mwa_hyperdrive::calibrate::di::code::read_vis_data
             at ./src/calibrate/di/code/mod.rs:370:9
  30: mwa_hyperdrive::calibrate::di::code::get_cal_vis::{{closure}}::{{closure}}
             at ./src/calibrate/di/code/mod.rs:185:26
  31: crossbeam_utils::thread::ScopedThreadBuilder::spawn::{{closure}}
             at /home/dev/.cargo/registry/src/github.com-1ecc6299db9ec823/crossbeam-utils-0.8.11/src/thread.rs:438:31
  32: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  33: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/alloc/src/boxed.rs:1951:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

This happens because in `vis_io::read::raw::RawDataReader::read_inner`, `data_array` is sized to take only the unflagged fine channels from the first fence, and a raw data reader is not aware of any info about fences.

And yes, I'm aware that I introduced this bug :P I just wanted to start a discussion about how to fix it, since it's now affecting users in production.

**Comment by d3v-null** [comment]
I pushed a potential fix to https://github.com/MWATelescope/mwa_hyperdrive/tree/issue-17 but it's not great, it just involves flagging all the channels outside of the first fence.

I don't think we'll be able to fix this without significant change to the API, so an easier fix would be to print out a more detailed warning about how to call hyperdrive with a subset of coarse channels, and not even attempt to calibrate multiple fences.

**Comment by cjordan** [comment]
This issue is fixed in 5cc5c40. The underlying issue, indeed, requires a bit more work.
